crypto/cmp/cmp_server.c

```diff
@@ -0,0 +1,656 @@
+        OSSL_CMP_log1(WARN, ctx, "ignoring %s protection of request message",
+                      invalid_protection ? "invalid" : "missing");
+        return 1;
+    }
+    if (req->body->type == OSSL_CMP_PKIBODY_ERROR && ctx->unprotectedErrors) {
+        OSSL_CMP_warn(ctx, "ignoring missing protection of error message");
+        return 1;
+    }
+    return 0;
+}
+
+/*
+ * Mocks the server/responder.
+ * srv_ctx is the context of the server
+ * returns 1 if a message was created and 0 on error
+ */
+static int process_request(OSSL_CMP_SRV_CTX *srv_ctx, OSSL_CMP_MSG *req,
+                           OSSL_CMP_MSG **rsp)
+{
+    cmp_srv_process_cb_t process_cb = NULL;
+    OSSL_CMP_CTX *ctx;
+
+    if (!ossl_assert(srv_ctx != NULL && srv_ctx->ctx != NULL && req != NULL
+                     && rsp != NULL))
+        return 0;
+
+    ctx = srv_ctx->ctx;
+    *rsp = NULL;
+
+    if (req->header->sender->type != GEN_DIRNAME) {
+        CMPerr(0, CMP_R_SENDER_GENERALNAME_TYPE_NOT_SUPPORTED);
+        return 0;
+    }
+    if (!X509_NAME_set(&ctx->recipient, req->header->sender->d.directoryName))
+        return 0;
+
+    if (ossl_cmp_msg_check_received(ctx, req, unprotected_exception,
+                                    srv_ctx->acceptUnprotectedRequests) < 0) {
+        CMPerr(0, CMP_R_FAILED_TO_RECEIVE_PKIMESSAGE);
+        return 0;
+    }
+    if (srv_ctx->sendError) {
+        if ((*rsp = ossl_cmp_error_new(ctx, srv_ctx->pkiStatusOut, -1, NULL,
+                                       srv_ctx->sendUnprotectedErrors)))
+            return 1;
+        CMPerr(0, CMP_R_ERROR_CREATING_ERROR);
+        return 0;
+    }
+
+    switch (req->body->type) {
+    case OSSL_CMP_PKIBODY_IR:
+        process_cb = srv_ctx->process_ir_cb;
+        break;
+    case OSSL_CMP_PKIBODY_CR:
+        process_cb = srv_ctx->process_cr_cb;
+        break;
+    case OSSL_CMP_PKIBODY_P10CR:
+        process_cb = srv_ctx->process_p10cr_cb;
+        break;
+    case OSSL_CMP_PKIBODY_KUR:
+        process_cb = srv_ctx->process_kur_cb;
+        break;
+    case OSSL_CMP_PKIBODY_POLLREQ:
+        process_cb = srv_ctx->process_pollreq_cb;
+        break;
+    case OSSL_CMP_PKIBODY_RR:
+        process_cb = srv_ctx->process_rr_cb;
+        break;
+    case OSSL_CMP_PKIBODY_ERROR:
+        process_cb = srv_ctx->process_error_cb;
+        break;
+    case OSSL_CMP_PKIBODY_CERTCONF:
+        process_cb = srv_ctx->process_certconf_cb;
+        break;
+    case OSSL_CMP_PKIBODY_GENM:
+        process_cb = srv_ctx->process_genm_cb;
+        break;
+    default:
+        CMPerr(0, CMP_R_UNEXPECTED_PKIBODY);
+        break;
+    }
+    if (process_cb == NULL)
+        return 0;
+    if ((*rsp = process_cb(srv_ctx, req)) == NULL)
+        return 0;
+
+    return 1;
+}
+
+/*
+ * Mocks the server connection. Works similar to OSSL_CMP_MSG_http_perform.
+ * A OSSL_CMP_SRV_CTX must be set as transfer_cb_arg
+ * returns 0 on success and else a CMP error reason code defined in cmp.h
```

Comment:
Hmmm! This is contrary to standard OpenSSL practice where 1 is success and 0 is an error. Why return a reason code and not just put an error on the stack?

Reply:
I cannot recall what the design decision was, nor even who did it (@mpeylo or me),
and I see no strong reason for it except maybe simplicity.
We can change that (also in `cmp_client.c`, which be the core of chunk 9).

Reply:
Fixed.

Reply:
@Akretsch, please just correct the comment as I indicated for the commit.

Reply:
@Akretsch, also the documentation needs to be update accordingly.

Reply:
... but for now this pertains only to our internal branch, `cmp-dev`.